artifact.go

```diff
@@ -203,67 +213,423 @@ func (m artifactModule) getBuildTestsResultsHandler(c *gin.Context) {
+	}
+
+	summaryList.Total = summaryList.Failed + summaryList.Passed + summaryList.Skipped
+
+	c.JSON(http.StatusOK, summaryList)
+}
+
+func parseMultipartFormData(c *gin.Context, buildID uint) ([]file, bool) {
+	files := make([]file, 0)
+
+	form, err := c.MultipartForm()
+	if err != nil {
+		ginutil.WriteMultipartFormReadError(c, err,
+			fmt.Sprintf("Failed reading multipart-form content from request body when uploading new"+
+				" artifact for build with ID %d.", buildID))
+		return nil, false
+	}
+
 	for k := range form.File {
 		if fhs := form.File[k]; len(fhs) > 0 {
-			f, err := fhs[0].Open()
-			if err != nil {
-				ginutil.WriteMultipartFormReadError(c, err, fmt.Sprintf(
-					"Failed with starting to read file content from multipart form request body when uploading new artifact for build with ID %d.",
-					buildID))
-				return
+			data, ok := readMultipartFileData(c, buildID, fhs[0])
+			if !ok {
+				return files, false
 			}
 
-			data, err := ioutil.ReadAll(f)
-			if err != nil {
-				ginutil.WriteMultipartFormReadError(c, err, fmt.Sprintf(
-					"Failed reading file content from multipart form request body when uploading new artifact for build with ID %d.",
-					buildID))
-				return
-			}
+			files = append(files, file{
+				name:     k,
+				fileName: fhs[0].Filename,
+				data:     data,
+			})
+		}
+	}
 
-			artifact := Artifact{
-				Data:     data,
-				Name:     k,
-				FileName: fhs[0].Filename,
-				BuildID:  buildID,
-			}
+	return files, true
+}
 
-			err = m.Database.Create(&artifact).Error
-			if err != nil {
-				ginutil.WriteDBWriteError(c, err, fmt.Sprintf(
-					"Failed saving artifact with name %q for build with ID %d in database.",
-					artifact.FileName, buildID))
-				return
-			}
+func readMultipartFileData(c *gin.Context, buildID uint, fh *multipart.FileHeader) ([]byte, bool) {
+	if fh == nil {
+		return nil, false
+	}
 
-			log.Info().
-				WithString("filename", artifact.Name).
-				WithUint("build", buildID).
-				WithUint("artifact", artifact.ArtifactID).
-				Message("File saved as artifact")
+	f, err := fh.Open()
+	if err != nil {
+		ginutil.WriteMultipartFormReadError(c, err,
+			fmt.Sprintf("Failed with starting to read file content from multipart form request body when"+
+				" uploading new artifact for build with ID %d.", buildID))
+		return nil, false
+	}
+	defer f.Close()
+
+	data, err := ioutil.ReadAll(f)
+	if err != nil {
+		ginutil.WriteMultipartFormReadError(c, err,
+			fmt.Sprintf("Failed reading file content from multipart form request body when uploading new"+
+				" artifact for build with ID %d.", buildID))
+		return nil, false
+	}
+	return data, true
+}
+
+type testRun struct {
+	XMLName       xml.Name      `xml:"TestRun"`
+	Results       results       `xml:"Results"`
+	ResultSummary resultSummary `xml:"ResultSummary"`
+}
+
+type results struct {
+	XMLName         xml.Name         `xml:"Results"`
+	UnitTestResults []unitTestResult `xml:"UnitTestResult"`
+}
```

Comment:
Some of these types has quite generic names. Especially the `results`, `output`, `message`, `stackTrace`, and `counters` ones. Could you prefix them with something like `xml` or `trx` to make it clearer where they're used, as well as to not produce unexpected behaviour from name collisions? Alternatively move these types to their own package, such as `/pkg/testresults/fbxtests` or something, together with the parsing

Also, you can nest anonymous structs in other structs. Maybe it's something you want to apply here?

```go
type trxTestRun struct {
	XMLName xml.Name `xml:"TestRun"`
	Results struct {
		XMLName         xml.Name `xml:"Results"`
		UnitTestResults []struct {
			XMLName   xml.Name `xml:"UnitTestResult"`
			TestName  string   `xml:"testName,attr"`
			Duration  string   `xml:"duration,attr"`
			StartTime string   `xml:"startTime,attr"`
			EndTime   string   `xml:"endTime,attr"`
			Outcome   string   `xml:"outcome,attr"`
			Output    output   `xml:"Output"`
		} `xml:"UnitTestResult"`
	} `xml:"Results"`
	ResultSummary resultSummary `xml:"ResultSummary"`
}
```

Doing so makes it extreamly difficult to create the subtypes separately without having to redeclare the entire type (such as in function parameters), but when you only work with the full object then this works really niceley